src/compiler/checker.ts

```diff
@@ -14441,11 +14442,26 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function getResolvedApparentTypeOfMappedType(type: MappedType) {
-        const typeVariable = getHomomorphicTypeVariable(type);
-        if (typeVariable && !type.declaration.nameType) {
-            const constraint = getConstraintOfTypeParameter(typeVariable);
+        if (contains(homomorphicMappedTypeApparentTypeStack, type)) {
+            return type;
+        }
+        const mappedType = type.target as MappedType || type;
+        const typeVariable = getHomomorphicTypeVariable(mappedType);
+        if (typeVariable && !mappedType.declaration.nameType) {
+            let constraint: Type | undefined;
+            if (!type.target) {
+                constraint = getConstraintOfTypeParameter(typeVariable);
+            }
+            else {
+                const modifiersConstraint = getConstraintOfType(getModifiersTypeFromMappedType(type));
+                if (modifiersConstraint) {
+                    homomorphicMappedTypeApparentTypeStack.push(type);
+                    constraint = getApparentType(modifiersConstraint);

```

Comment:
I'm not super sure if I should be calling `getApparentType` here but then... we get to this place from `getApparentType` so it feels like it's OK to do right now.